### PR TITLE
fix(quire): size cfloat quire radix_point from smallest normal LSB (#1202)

### LINKED ISSUE #1202: quire_traits<cfloat> radix_point undersized for no-subnormals configs -> quire dot not exact
## Summary

For a `cfloat` **without subnormals** (the default), `quire_traits<cfloat>`
sizes the quire's fraction field (`radix_point`) too small: it reserves room
for `minpos^2` but **not** for the product of a *normal* small value, whose
significand extends `fbits` bits below its scale. Products of small-but-normal
operands are therefore truncated in the quire, so an exact fused dot floors at
roughly `2^-radix_point` instead of being exact.

## Evidence

Dot of 4096 pseudo-random pairs in (-1,1), accumulated via `quire<T>` +
`quire_mul`, delivered with `convert_to<double>()`, relative error vs a
long-double reference over the same quantized values:

| type                              | radix_point | dot rel. error |
|-----------------------------------|------------:|---------------:|
| `cfloat<16,5>` (default, no subn.)|          28 |  **4.6e-10**   |
| `cfloat<16,5,uint16_t,true,..>` (subnormals on) | 48 |  **0** (exact) |
| `cfloat<32,8>` (default)          |         252 |   1.1e-16      |

The **single-product** `quire_mul(cfloat<16,5>, cfloat<16,5>)` is already exact
(verified: product error 0) — the loss is purely in the quire's fraction width,
and turning subnormals on (which enlarges `radix_point` to 48) makes the whole
dot exact. `4.6e-10 ≈ 2^-31`, consistent with a fraction field ~`2*fbits`
short of what a normal-operand product needs.

## Root cause

`include/sw/universal/traits/quire_traits.hpp`, `quire_traits<cfloat>`:

```cpp
static constexpr unsigned abs_min_scale = hasSubnormals
    ? (bias + fbits - 1u)     // subnormal path already includes fbits
    : (bias >= 1u ? bias - 1u : 0u);   // <-- normal path omits fbits
static constexpr unsigned radix_point = 2u * abs_min_scale;  // minpos^2 at bit 0
```

In the no-subnormals branch `abs_min_scale` is the scale of `minpos` (a single
significand bit), so `radix_point = 2*(bias-1)`. But the smallest *normal*
value is `2^(1-bias)` with `fbits` fraction bits — its LSB sits at scale
`(1-bias-fbits)`, and its square needs `2*(bias-1) + 2*fbits` fraction bits.
The subnormal branch happens to be large enough only because `min_scale` there
already descends by `fbits`.

## Requested fix

Size `radix_point` from the smallest *normal* product's LSB in the
no-subnormals case, e.g.

```cpp
static constexpr unsigned radix_point = 2u * (abs_min_scale + fbits);
```

(or otherwise guarantee `radix_point >= 2*(abs_min_scale) + product_fbits`).
Please add a regression check that an exact `quire` dot over near-minpos normal
operands is bit-exact for small cfloat configs (`cfloat<8,2>`, `cfloat<16,5>`)
in both subnormal modes.

## Impact

Makes the cfloat quire non-exact for the common no-subnormals configs — a
`cfloat<16,5>` "exact" dot is actually ~1e-9, *worse* than simply accumulating
the same products in `double`. Surfaced in the mp-blas level-1 accumulator study.

### PR DESCRIPTION
**Pull request by Ravenwater**
## Summary
`quire_traits<cfloat>::radix_point` was undersized for **no-subnormals** cfloat
configs (the default), so fused dot products of small-but-normal operands were
truncated -- a `cfloat<16,5>` "exact" dot was ~1e-9 (worse than `double`), and
`cfloat<8,2>` had `radix_point == 0` (no fraction bits at all).

## Root cause
The no-subnormals branch derived `radix_point` from `|min value scale|`
(`bias-1`) -- the scale of the smallest normal's **hidden bit**. But that value
carries `fbits` fraction bits below it, down to `2^(1-bias-fbits)`, so its
square needs `2*(bias+fbits-1)` fraction bits. The old value `2*(bias-1)` was
`2*fbits` short.

The smallest representable magnitude's LSB is at scale `1-bias-fbits` in **both**
subnormal modes (without subnormals the smallest normal's fraction reaches
exactly as low as the smallest subnormal), so `radix_point` does not depend on
`hasSubnormals`.

## Fix
`abs_min_scale = bias + fbits - 1` (guarded for the degenerate `bias+fbits==0`
case), unifying both branches. This matches:
- the native `float`/`double` `quire_traits` specializations (which already use
  `bias + fbits - 1` unconditionally), and
- the file's own documented examples (`cfloat<8,3>=12`, `cfloat<32,8>=298`) --
  which the buggy code contradicted (it produced 4 and 252).

| config | old radix_point | new (this PR) |
|---|--:|--:|
| cfloat<16,5> no-sub | 28 | 48 |
| cfloat<8,2> no-sub  | 0  | 10 |
| cfloat<32,8> no-sub | 252 | 298 |

## Regression
`static/quire/api/cfloat.cpp::TestExactDotSmallNormals`: an exact quire dot over
near-minpos **normal** operands, delivered via `quire::convert_to<double>()`,
must be bit-exact vs an independent `double` oracle for `cfloat<8,2>` and
`cfloat<16,5>` in **both** subnormal modes; plus `static_assert`s on the
documented `radix_point` values. Fails on pre-fix headers (cfloat<8,2> delivered
8 vs exact 10.37; cfloat<16,5> ~1e-9 off), passes with the fix.

## Test Results (gcc + clang)
| Target | Result |
|--------|--------|
| quire_traits / cfloat / cfloat_fdp | PASS |
| aggregator_fdp / api / fixpnt_fdp / lns_fdp | PASS |
| new TestExactDotSmallNormals (both subnormal modes) | PASS |

The larger no-subnormals cfloat quire introduces no regressions (fixpnt/lns
specializations untouched).

## Test plan
- [ ] Fast CI passes (gcc + clang CI_LITE)
- [ ] Promote to ready when satisfied

Resolves #1202

Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Corrected quire scaling for `cfloat` values, including consistent handling of fraction bits and zero-scale edge cases.
  * Ensured radix-point calculations remain consistent whether subnormal values are enabled or disabled.

* **Tests**
  * Added coverage verifying exact fused dot-product results for values near the minimum normal range.
  * Confirmed consistent behavior across both subnormal modes.

### CODE REVIEW

File: include/sw/universal/traits/quire_traits.hpp [hunk 1]
```diff
@@ -110,15 +117,14 @@ struct quire_traits<cfloat<nbits, es, bt, hasSubnormals, hasMaxExpValues, isSatu
...
+	// subnormal modes (see the note above; #1202). Matches the native float/double
+	// specializations below, which use bias + fbits - 1 unconditionally. The guard covers
+	// the degenerate bias + fbits == 0 case (e.g. cfloat<2,1>).
+	static constexpr unsigned abs_min_scale = (bias + fbits >= 1u) ? (bias + fbits - 1u) : 0u;
 
 	// Quire geometry derived from product scale bounds
-	static constexpr unsigned radix_point = 2u * abs_min_scale;   // minpos^2 at bit 0
+	static constexpr unsigned radix_point = 2u * abs_min_scale;   // smallest product bit at bit 0
 	static constexpr unsigned upper_range = 2u * max_scale + 2u;  // maxpos^2 MSB at bit range-1
 	static constexpr unsigned range       = radix_point + upper_range;
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/sw/universal/traits/quire_traits.hpp:129]: Fixed in cd33c2b8. Those geometry comments had actually drifted before this PR too (radix_point on main was 252, not the documented 281), and this PR moves it to 298 / qbits 584. Rather than re-derive the fragile per-limb scale->boundary table (which keeps going stale), I replaced both header blocks with a concise correct summary (radix_point=298, upper_range=256, range=554, qbits=584; base_offset=252+scale) pointing to `quire_traits<>` as the source of truth, and dropped the table. Thanks for confirming no other `hasSubnormals=false` sizing constants/assertions remain.
